File: src/com/cattailsw/nanidroid/IncomingNarIntent.java
package com.cattailsw.nanidroid;

import android.content.Intent;
import android.net.Uri;

/**
 * The only externally accepted install/update entry point until the SAF picker
 * is introduced.  In particular, do not turn a caller-controlled file:// URI
 * into a private filesystem path.
 */
final class IncomingNarIntent {
    private IncomingNarIntent() {
    }

    static boolean isApprovedDownload(Intent intent) {
        return intent != null
                && Intent.ACTION_VIEW.equals(intent.getAction())
                && isApprovedDownload(intent.getData());
    }

    static boolean isApprovedDownload(Uri uri) {
        if (uri == null || !"https".equalsIgnoreCase(uri.getScheme())) {
            return false;
        }
        if (uri.getHost() == null || uri.getHost().length() == 0) {
            return false;
        }
        String path = uri.getPath();
        if (path == null) {
            return false;
        }
        String lowerPath = path.toLowerCase(java.util.Locale.US);
        return lowerPath.endsWith(".nar") || lowerPath.endsWith(".zip");
    }
}

File: src/main/kotlin/com/cattailsw/nanidroid/IncomingNarIntent.kt
package com.cattailsw.nanidroid

import android.content.Intent
import android.net.Uri
import java.util.Locale

/**
 * The only externally accepted install/update entry point until the SAF picker
 * is introduced. Caller-controlled file URIs never become private paths.
 */
object IncomingNarIntent {
    @JvmStatic
    fun isApprovedDownload(intent: Intent?): Boolean =
        Intent.ACTION_VIEW == intent?.action && isApprovedDownload(intent.data)

    @JvmStatic
    fun isApprovedDownload(uri: Uri?): Boolean {
        val download = uri ?: return false
        if (!download.scheme.equals("https", ignoreCase = true)) return false
        if (download.host.isNullOrEmpty()) return false
        val path = download.path ?: return false
        val lowerPath = path.lowercase(Locale.US)
        return lowerPath.endsWith(".nar") || lowerPath.endsWith(".zip")
    }
}
